File: src/camera/capture.py
"""Webcam capture and frame streaming."""

from __future__ import annotations

from pathlib import Path

import cv2
import numpy as np

from src.calibration.matrices import load_calibration_matrices
from src.camera.discovery import discover_capture, list_devices
from src.camera.undistort import FrameUndistorter
from src.config import load_section, resolve_config_path
from src.exceptions import CameraError


class ChessCamera:
    """Manages a USB webcam stream with configurable resolution and undistortion."""

    def __init__(self, config_path: str | Path | None = None) -> None:
        self._config_path = resolve_config_path(config_path)
        self._config = load_section("camera", self._config_path)
        self._cap: cv2.VideoCapture | None = None
        self._opened = False
        self._active_device: int | str | None = None
        self._undistorter: FrameUndistorter | None = None
        self._load_undistorter()

    def _load_undistorter(self) -> None:
        matrices = load_calibration_matrices(self._config_path)
        if matrices is None:
            return
        camera_matrix, dist_coeffs = matrices
        self._undistorter = FrameUndistorter(camera_matrix, dist_coeffs)

    @property
    def device_id(self) -> int:
        return int(self._config.get("device_id", 0))

    @property
    def device_path(self) -> str | None:
        path = self._config.get("device_path")
        return None if path in (None, "null", "") else str(path)

    @property
    def width(self) -> int:
        return int(self._config.get("width", 1280))

    @property
    def height(self) -> int:
        return int(self._config.get("height", 720))

    @property
    def fps(self) -> int:
        return int(self._config.get("fps", 30))

    @property
    def is_open(self) -> bool:
        return self._opened and self._cap is not None and self._cap.isOpened()

    @property
    def is_calibrated(self) -> bool:
        return self._undistorter is not None

    @property
    def active_device(self) -> int | str | None:
        return self._active_device

    @staticmethod
    def list_devices(max_index: int = 10) -> list[dict]:
        return list_devices(max_index)

    def open(self) -> None:
        if self.is_open:
            return

        discovered = discover_capture(self.device_path, self.device_id)
        if discovered is None:
            available = self.list_devices()
            names = [device["path"] for device in available] or ["none"]
            raise CameraError(
                "Failed to open any camera device. "
                "Check that the webcam is connected, not in use by another app, "
                "and that your user is in the 'video' group. "
                f"Set camera.device_path or camera.device_id in config/camera_config.yaml. "
                f"Working devices found: {', '.join(names)}."
            )

        device, cap = discovered
        self._active_device = device

        configured = self.device_path or self.device_id
        if device != configured:
            print(f"Camera: using {device} (configured {configured} unavailable)")

        cap.set(cv2.CAP_PROP_FRAME_WIDTH, self.width)
        cap.set(cv2.CAP_PROP_FRAME_HEIGHT, self.height)
        cap.set(cv2.CAP_PROP_FPS, self.fps)

        self._cap = cap
        self._opened = True
        if self._undistorter is not None:
            self._undistorter.reset()

    def read(self, apply_undistort: bool = True) -> np.ndarray:
        if not self.is_open:
            raise CameraError("Camera is not open. Call open() first.")

        assert self._cap is not None
        ok, frame = self._cap.read()
        if not ok or frame is None:
            raise CameraError(
                "Failed to read frame from camera. "
                "The device may have been disconnected."
            )

        if apply_undistort and self._undistorter is not None:
            return self._undistorter.apply(frame)

        return frame

    def release(self) -> None:
        if self._cap is not None:
            self._cap.release()
            self._cap = None
        self._opened = False
        self._active_device = None
        if self._undistorter is not None:
            self._undistorter.reset()

    def __enter__(self) -> ChessCamera:
        self.open()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.release()

File: src/camera/capture.cpp
#include "chess_robot/camera/capture.hpp"

#include <iostream>
#include <sstream>

#include "chess_robot/calibration/matrices.hpp"
#include "chess_robot/config.hpp"
#include "chess_robot/exceptions.hpp"

namespace chess_robot {

ChessCamera::ChessCamera(const std::optional<fs::path>& config_path)
    : config_path_(resolveConfigPath(config_path)) {
  YAML::Node camera = loadSection("camera", config_path_);
  device_id_ = camera["device_id"] ? camera["device_id"].as<int>() : 0;
  if (camera["device_path"] && !camera["device_path"].IsNull()) {
    const std::string path = camera["device_path"].as<std::string>();
    if (!path.empty() && path != "null") {
      device_path_ = path;
    }
  }
  width_ = camera["width"] ? camera["width"].as<int>() : 1280;
  height_ = camera["height"] ? camera["height"].as<int>() : 720;
  fps_ = camera["fps"] ? camera["fps"].as<int>() : 30;
  loadUndistorter();
}

void ChessCamera::loadUndistorter() {
  auto matrices = loadCalibrationMatrices(config_path_);
  if (!matrices.has_value()) {
    return;
  }
  undistorter_ = std::make_unique<FrameUndistorter>(
      matrices->camera_matrix, matrices->dist_coeffs, matrices->image_size);
}

bool ChessCamera::isOpen() const {
  return opened_ && cap_.isOpened();
}

std::vector<DeviceInfo> ChessCamera::listDevices(int max_index) {
  return chess_robot::listDevices(max_index);
}

void ChessCamera::open() {
  if (isOpen()) {
    return;
  }

  auto discovered = discoverCapture(device_path_, device_id_);
  if (!discovered.has_value()) {
    const auto available = listDevices();
    std::ostringstream names;
    if (available.empty()) {
      names << "none";
    } else {
      for (std::size_t i = 0; i < available.size(); ++i) {
        if (i > 0) {
          names << ", ";
        }
        names << available[i].path;
      }
    }
    throw CameraError(
        "Failed to open any camera device. "
        "Check that the webcam is connected, not in use by another app, "
        "and that your user is in the 'video' group. "
        "Set camera.device_path or camera.device_id in config/camera_config.yaml. "
        "Working devices found: " +
        names.str() + ".");
  }

  active_device_ = discovered->device_label;
  const std::string configured =
      device_path_.has_value() ? *device_path_ : std::to_string(device_id_);
  if (*active_device_ != configured) {
    std::cout << "Camera: using " << *active_device_ << " (configured " << configured
              << " unavailable)\n";
  }

  cap_ = std::move(discovered->capture);
  cap_.set(cv::CAP_PROP_FRAME_WIDTH, width_);
  cap_.set(cv::CAP_PROP_FRAME_HEIGHT, height_);
  cap_.set(cv::CAP_PROP_FPS, fps_);
  opened_ = true;
  if (undistorter_) {
    undistorter_->reset();
  }
}

cv::Mat ChessCamera::read(bool apply_undistort) {
  if (!isOpen()) {
    throw CameraError("Camera is not open. Call open() first.");
  }

  cv::Mat frame;
  if (!cap_.read(frame) || frame.empty()) {
    throw CameraError(
        "Failed to read frame from camera. "
        "The device may have been disconnected.");
  }

  if (apply_undistort && undistorter_) {
    return undistorter_->apply(frame);
  }
  return frame;
}

void ChessCamera::release() {
  if (cap_.isOpened()) {
    cap_.release();
  }
  opened_ = false;
  active_device_.reset();
  if (undistorter_) {
    undistorter_->reset();
  }
}

}  // namespace chess_robot
